### PR TITLE
New task query functionality plugin

### PR DESCRIPTION
**Pull request by indvamshi**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RWA-744

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java [hunk 1]
```diff
@@ -23,7 +26,7 @@ public CftQueryService(TaskResourceRepository taskResourceRepository) {
         this.taskResourceRepository = taskResourceRepository;
     }
 
-    public List<TaskResource> getAllTasks(
+    public GetTasksResponse<Task> getAllTasks(
```
Comment (mkayad) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java:29]: wouldn't be better to keep List<TaskResource> at this level and not use the GetTasksResponse which is the upper level?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java [hunk 2]
```diff
@@ -36,9 +39,29 @@ public CftQueryService(TaskResourceRepository taskResourceRepository) {
...
-        final Page<TaskResource> taskResources = taskResourceRepository.findAll(taskResourceSpecification, page);
+        final Page<TaskResource> pages = taskResourceRepository.findAll(taskResourceSpecification, page);
 
-        return taskResources.toList();
+        final List<TaskResource> taskResources = pages.toList();
+
+        return mapToTask(taskResources, pages.getTotalElements());
+    }
+
+    private GetTasksResponse<Task> mapToTask(List<TaskResource> taskResources, long totalNumberOfTasks) {
```
Comment (mkayad) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java:49]: Same thing here, why not keep the standard list and wrap the answer in GetTasksResponse in the controller

**Review by mkayad — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java (hunk 1 above)
Comment (indvamshi) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java:29]: following the standards , making them consistent across the apis.

Comment (mkayad) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java:29]: Which standards are you referring to? i think class dependency should be downward not upward, it should not depend on class which are in the controllers package. We ask other devs if you want.

Comment (indvamshi) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java:29]: I got it. will change.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java (hunk 2 above)
Comment (indvamshi) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java:49]: I got it. will change.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java (hunk 1 above)
Comment (indvamshi) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java:29]: The reason being the totalNumber of records cannot be added separately in controller. For MVP , to get total number or records extra call made to get total number of records but in jpa case its overkill.

**Review by mkayad — DISMISSED** [review]
LGTM

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java [hunk 3]
```diff
@@ -36,9 +39,29 @@ public CftQueryService(TaskResourceRepository taskResourceRepository) {
...
+        final List<Task> tasks = taskResources.stream().map(taskResource ->
+            new Task(taskResource.getTaskId(), taskResource.getTaskName(), taskResource.getTaskType(),
+                taskResource.getState().getValue(), taskResource.getTaskSystem().getValue(),
+                taskResource.getSecurityClassification().getSecurityClassification(),
+                taskResource.getTitle(),
+                taskResource.getCreated() == null ? null : taskResource.getCreated().toZonedDateTime(),
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java:55]: can we add unit tests to cover null scenarios for these 2 lines?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java [hunk 4]
```diff
@@ -163,19 +173,21 @@ private static Predicate mapSecurityClassification(
                     SecurityClassification.PUBLIC
                 )
             );
-        } else {
+        } else if (classification.equals(Classification.PUBLIC)) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java:176]: unit test for when !equal?

Comment (indvamshi) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java:176]: I have to move that else part, classification cannot be other than private, public and restricted.
